ddl/index.go

```diff
@@ -1232,6 +1254,136 @@ func indexColumnSliceEqual(a, b []*model.IndexColumn) bool {
 	return true
 }
 
+type cleanUpIndexWorker struct {
+	baseIndexWorker
+}
+
+func newCleanUpIndexWorker(sessCtx sessionctx.Context, worker *worker, id int, t table.PhysicalTable, decodeColMap map[int64]decoder.Column, sqlMode mysql.SQLMode) *cleanUpIndexWorker {
+	indexes := make([]table.Index, 0, len(t.Indices()))
+	rowDecoder := decoder.NewRowDecoder(t, t.WritableCols(), decodeColMap)
+	for _, index := range t.Indices() {
+		if index.Meta().Global {
+			indexes = append(indexes, index)
+		}
+	}
+	return &cleanUpIndexWorker{
+		baseIndexWorker: baseIndexWorker{
+			backfillWorker: newBackfillWorker(sessCtx, worker, id, t),
+			indexes:        indexes,
+			rowDecoder:     rowDecoder,
+			defaultVals:    make([]types.Datum, len(t.WritableCols())),
+			rowMap:         make(map[int64]types.Datum, len(decodeColMap)),
+			metricCounter:  metrics.BackfillTotalCounter.WithLabelValues("add_idx_speed"),
+			sqlMode:        sqlMode,
+		},
+	}
+}
+
+func (w *cleanUpIndexWorker) BackfillDataInTxn(handleRange reorgBackfillTask) (taskCtx backfillTaskContext, errInTxn error) {
+	failpoint.Inject("errorMockPanic", func(val failpoint.Value) {
+		if val.(bool) {
+			panic("panic test")
+		}
+	})
+
+	oprStartTime := time.Now()
+	errInTxn = kv.RunInNewTxn(w.sessCtx.GetStore(), true, func(txn kv.Transaction) error {
+		taskCtx.addedCount = 0
+		taskCtx.scanCount = 0
+		txn.SetOption(kv.Priority, w.priority)
+
+		idxRecords, nextHandle, taskDone, err := w.fetchRowColVals(txn, handleRange)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		taskCtx.nextHandle = nextHandle
+		taskCtx.done = taskDone
+
+		n := len(w.indexes)
+		for i, idxRecord := range idxRecords {
+			taskCtx.scanCount++
+			// we fetch records row by row, so records will belong to
+			// index[0], index[1] ... index[n-1], index[0], index[1] ...
+			// respectively. So indexes[i%n] is the index of idxRecords[i].
+			err := w.indexes[i%n].Delete(w.sessCtx.GetSessionVars().StmtCtx, txn, idxRecord.vals, idxRecord.handle)
+			if err != nil {
+				return errors.Trace(err)
+			}
+			taskCtx.addedCount++
```

Comment:
Maybe we should replace all "backfilling" in `backfilling.go` to "reorganize", because cleaning up index is not "backfilling", but it a big project, so I leave it unchanged now.